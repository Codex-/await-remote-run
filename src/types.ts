/**
 * The Status and Conclusion types are difficult to find a reliable source
 * of truth for, but this seems accurate from testing:
 * https://docs.github.com/en/enterprise-server@3.14/rest/guides/using-the-rest-api-to-interact-with-checks#about-check-suites
 */

export enum WorkflowRunStatus {
  Queued = "queued",
  InProgress = "in_progress",
  Requested = "requested",
  Waiting = "waiting",
  Pending = "pending",
  Completed = "completed",
}

export enum WorkflowRunConclusion {
  Success = "success",
  Failure = "failure",
  Neutral = "neutral",
  Cancelled = "cancelled",
  Skipped = "skipped",
  TimedOut = "timed_out",
  Stale = "stale",
  StartupFailure = "startup_failure",
  ActionRequired = "action_required",
}

export type Result<T> = ResultSuccess<T> | RequestTimeout | ResultUnsupported;

interface ResultSuccess<T> {
  success: true;
  value: T;
}

interface RequestTimeout {
  success: false;
  reason: "timeout";
}

interface ResultUnsupported {
  success: false;
  reason: "unsupported";
  value: string;
}

export type WorkflowRunStatusResult =
  | ResultSuccess<WorkflowRunStatus>
  | ResultStatusPending
  | ResultUnsupported;

interface ResultStatusPending {
  success: false;
  reason: "pending";
  value: WorkflowRunStatus;
}

export type WorkflowRunConclusionResult =
  | ResultSuccess<WorkflowRunConclusion>
  | ResultConclusionInconclusive
  | ResultUnsupported;

interface ResultConclusionInconclusive {
  success: false;
  reason: "inconclusive" | "timeout";
  value: WorkflowRunConclusion;
}
