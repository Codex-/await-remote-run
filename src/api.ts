import * as core from "@actions/core";
import * as github from "@actions/github";
import type { GitHub } from "@actions/github/lib/utils";
import { ActionConfig, getConfig } from "./action";

type Octokit = InstanceType<typeof GitHub>;

let config: ActionConfig;
let octokit: Octokit;

export enum WorkflowRunStatus {
  Queued = "queued",
  InProgress = "in_progress",
  Completed = "completed",
}

export enum WorkflowRunConclusion {
  Success = "success",
  Failure = "failure",
  Neutral = "neutral",
  Cancelled = "cancelled",
  Skipped = "skipped",
  TimedOut = "timed_out",
  ActionRequired = "action_required",
}

export function init(cfg?: ActionConfig): void {
  config = cfg || getConfig();
  octokit = github.getOctokit(config.token);
}

export interface WorkflowRunState {
  status: WorkflowRunStatus | null;
  conclusion: WorkflowRunConclusion | null;
}

export async function getWorkflowRunState(
  runId: number
): Promise<WorkflowRunState> {
  try {
    // https://docs.github.com/en/rest/reference/actions#get-a-workflow-run
    const response = await octokit.rest.actions.getWorkflowRun({
      owner: config.owner,
      repo: config.repo,
      run_id: runId,
    });

    if (response.status !== 200) {
      throw new Error(
        `Failed to get Workflow Run state, expected 200 but received ${response.status}`
      );
    }

    core.debug(
      `Fetched Run:\n` +
        `  Repository: ${config.owner}/${config.repo}\n` +
        `  Run ID: ${runId}\n` +
        `  Status: ${response.data.status}\n` +
        `  Conclusion: ${response.data.conclusion}`
    );

    return {
      status: response.data.status as WorkflowRunStatus | null,
      conclusion: response.data.conclusion as WorkflowRunConclusion | null,
    };
  } catch (error) {
    if (error instanceof Error) {
      core.error(
        `getWorkflowRunState: An unexpected error has occurred: ${error.message}`
      );
      error.stack && core.debug(error.stack);
    }
    throw error;
  }
}

export interface WorkflowRunJob {
  id: number;
  name: string;
  status: "queued" | "in_progress" | "completed";
  conclusion: string | null;
  steps: WorkflowRunJobStep[];
  url: string | null;
}

export interface WorkflowRunJobStep {
  name: string;
  status: string;
  conclusion: string | null;
  number: number;
}

type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;
type ListJobsForWorkflowRunResponse = Awaited<
  ReturnType<Octokit["rest"]["actions"]["listJobsForWorkflowRun"]>
>;

async function getWorkflowRunJobs(
  runId: number
): Promise<ListJobsForWorkflowRunResponse> {
  // https://docs.github.com/en/rest/reference/actions#list-jobs-for-a-workflow-run
  const response = await octokit.rest.actions.listJobsForWorkflowRun({
    owner: config.owner,
    repo: config.repo,
    run_id: runId,
    filter: "latest",
  });

  if (response.status !== 200) {
    throw new Error(
      `Failed to get Jobs for Workflow Run, expected 200 but received ${response.status}`
    );
  }

  return response;
}

export async function getWorkflowRunFailedJobs(
  runId: number
): Promise<WorkflowRunJob[]> {
  try {
    const response = await getWorkflowRunJobs(runId);
    const fetchedFailedJobs = response.data.jobs.filter(
      (job) => job.conclusion === "failure"
    );

    if (fetchedFailedJobs.length <= 0) {
      core.warning(`Failed to find failed Jobs for Workflow Run ${runId}`);
      return [];
    }

    const jobs: WorkflowRunJob[] = fetchedFailedJobs.map((job) => {
      const steps = job.steps?.map((step) => ({
        name: step.name,
        status: step.status,
        conclusion: step.conclusion,
        number: step.number,
      }));

      return {
        id: job.id,
        name: job.name,
        status: job.status,
        conclusion: job.conclusion,
        steps: steps || [],
        url: job.html_url,
      };
    });

    core.debug(
      `Fetched Jobs for Run:\n` +
        `  Repository: ${config.owner}/${config.repo}\n` +
        `  Run ID: ${config.runId}\n` +
        `  Jobs: [${jobs.map((job) => job.name)}]`
    );

    for (const job of jobs) {
      const steps = job.steps.map((step) => `${step.number}: ${step.name}`);
      core.debug(
        `  Job: ${job.name}\n` +
          `    ID: ${job.id}\n` +
          `    Status: ${job.status}\n` +
          `    Conclusion: ${job.conclusion}\n` +
          `    Steps: [${steps}]`
      );
    }

    return jobs;
  } catch (error) {
    if (error instanceof Error) {
      core.error(
        `getWorkflowRunJobFailures: An unexpected error has occurred: ${error.message}`
      );
      error.stack && core.debug(error.stack);
    }
    throw error;
  }
}

export async function getWorkflowRunActiveJobUrl(
  runId: number
): Promise<string> {
  try {
    const response = await getWorkflowRunJobs(runId);
    const fetchedInProgressJobs = response.data.jobs.filter(
      (job) => job.status === "in_progress"
    );

    if (fetchedInProgressJobs.length <= 0) {
      core.warning(`Failed to find in_progress Jobs for Workflow Run ${runId}`);
      return "Unable to fetch URL";
    }

    core.debug(
      `Fetched Jobs for Run:\n` +
        `  Repository: ${config.owner}/${config.repo}\n` +
        `  Run ID: ${config.runId}\n` +
        `  Jobs (in_progress): [${fetchedInProgressJobs.map(
          (job) => job.name
        )}]`
    );

    return (
      fetchedInProgressJobs[0].html_url || "GitHub failed to return the URL"
    );
  } catch (error) {
    if (error instanceof Error) {
      core.error(
        `getWorkflowRunActiveJobUrl: An unexpected error has occurred: ${error.message}`
      );
      error.stack && core.debug(error.stack);
    }
    throw error;
  }
}
