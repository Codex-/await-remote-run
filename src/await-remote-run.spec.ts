import * as core from "@actions/core";
import {
  afterAll,
  afterEach,
  beforeEach,
  describe,
  expect,
  it,
  vi,
  type MockInstance,
} from "vitest";

import * as api from "./api.ts";
import {
  getWorkflowRunConclusionResult,
  getWorkflowRunStatusResult,
  handleActionFail,
} from "./await-remote-run.ts";
import { mockLoggingFunctions } from "./test-utils/logging.mock.ts";
import { WorkflowRunConclusion, WorkflowRunStatus } from "./types.ts";

vi.mock("@actions/core");
vi.mock("@actions/github");
vi.mock("./api.ts");

describe("await-remote-run", () => {
  const {
    coreErrorLogMock,
    coreInfoLogMock,
    coreDebugLogMock,
    assertOnlyCalled,
    assertNoneCalled,
  } = mockLoggingFunctions();

  afterAll(() => {
    vi.restoreAllMocks();
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  describe("getWorkflowRunStatusResult", () => {
    it("should return success on completed status", () => {
      // Behaviour
      const result = getWorkflowRunStatusResult(WorkflowRunStatus.Completed, 0);
      if (!result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(true);
      expect(result.value).toStrictEqual(WorkflowRunStatus.Completed);

      // Logging
      assertNoneCalled();
    });

    it("should return inconclusive on queued status", () => {
      // Behaviour
      const result = getWorkflowRunStatusResult(WorkflowRunStatus.Queued, 0);
      if (result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(false);
      expect(result.reason).toStrictEqual("pending");

      // Logging
      assertOnlyCalled(coreDebugLogMock);
      expect(coreDebugLogMock).toHaveBeenCalledOnce();
      expect(coreDebugLogMock.mock.lastCall?.[0]).toMatchInlineSnapshot(
        `"Run is queued to begin, attempt 0..."`,
      );
    });

    it("should return inconclusive on in_progress status", () => {
      // Behaviour
      const result = getWorkflowRunStatusResult(
        WorkflowRunStatus.InProgress,
        0,
      );
      if (result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(false);
      expect(result.reason).toStrictEqual("pending");

      // Logging
      assertOnlyCalled(coreDebugLogMock);
      expect(coreDebugLogMock).toHaveBeenCalledOnce();
      expect(coreDebugLogMock.mock.lastCall?.[0]).toMatchInlineSnapshot(
        `"Run is in progress, attempt 0..."`,
      );
    });

    it.each([
      WorkflowRunStatus.Pending,
      WorkflowRunStatus.Requested,
      WorkflowRunStatus.Waiting,
    ])("should return unsupported on %s status", (status) => {
      // Behaviour
      const result = getWorkflowRunStatusResult(status, 0);
      if (result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(false);
      expect(result.reason).toStrictEqual("unsupported");

      // Logging
      assertOnlyCalled(coreDebugLogMock);
      expect(coreDebugLogMock).toHaveBeenCalledOnce();
      expect(coreDebugLogMock.mock.lastCall?.[0]).toStrictEqual(
        `Run status is unsupported: ${status}`,
      );
    });
  });

  describe("getWorkflowRunConclusionResult", () => {
    it("should return success on success conclusion", () => {
      // Behaviour
      const result = getWorkflowRunConclusionResult(
        WorkflowRunConclusion.Success,
      );
      if (!result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(true);
      expect(result.value).toStrictEqual(WorkflowRunConclusion.Success);

      // Logging
      assertNoneCalled();
    });

    it("should return non-success on an unsupported conclusion", () => {
      // Behaviour
      const result = getWorkflowRunConclusionResult(
        "random_conclusion" as WorkflowRunConclusion,
      );
      if (result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(false);
      expect(result.reason).toStrictEqual("unsupported");

      // Logging
      assertOnlyCalled(coreErrorLogMock, coreInfoLogMock);
      expect(coreErrorLogMock).toHaveBeenCalledOnce();
      expect(coreErrorLogMock.mock.lastCall?.[0]).toMatchInlineSnapshot(
        `"Run has failed with unsupported conclusion: random_conclusion"`,
      );
      expect(coreInfoLogMock).toHaveBeenCalledOnce();
      expect(coreInfoLogMock.mock.lastCall?.[0]).toMatchInlineSnapshot(
        `"Please open an issue with this conclusion value"`,
      );
    });

    it("should return non-success on timeout conclusion", () => {
      // Behaviour
      const result = getWorkflowRunConclusionResult(
        WorkflowRunConclusion.TimedOut,
      );
      if (result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(false);
      expect(result.reason).toStrictEqual("timeout");

      // Logging
      assertOnlyCalled(coreErrorLogMock);
      expect(coreErrorLogMock).toHaveBeenCalledOnce();
      expect(coreErrorLogMock.mock.lastCall?.[0]).toMatchInlineSnapshot(
        `"Run has timeout out"`,
      );
    });

    it.each([
      WorkflowRunConclusion.ActionRequired,
      WorkflowRunConclusion.Cancelled,
      WorkflowRunConclusion.Failure,
      WorkflowRunConclusion.Neutral,
      WorkflowRunConclusion.Skipped,
    ])("should return non-success on %s conclusion", (conclusion) => {
      // Behaviour
      const result = getWorkflowRunConclusionResult(conclusion);
      if (result.success) {
        expect.fail();
      }
      expect(result.success).toStrictEqual(false);
      expect(result.reason).toStrictEqual("inconclusive");

      // Logging
      assertOnlyCalled(coreErrorLogMock);
      expect(coreErrorLogMock).toHaveBeenCalledOnce();
      expect(coreErrorLogMock.mock.lastCall?.[0]).toStrictEqual(
        `Run has failed with conclusion: ${conclusion}`,
      );
    });
  });

  describe("handleActionFail", () => {
    let setFailedSpy: MockInstance<typeof core.setFailed>;
    let setOutputSpy: MockInstance<typeof core.setOutput>;

    let apiFetchWorkflowRunFailedJobsMock: MockInstance<
      typeof api.fetchWorkflowRunFailedJobs
    >;

    beforeEach(() => {
      setFailedSpy = vi.spyOn(core, "setFailed");
      setOutputSpy = vi.spyOn(core, "setOutput");

      apiFetchWorkflowRunFailedJobsMock = vi.spyOn(
        api,
        "fetchWorkflowRunFailedJobs",
      );
    });

    it("should set the action output and status", async () => {
      apiFetchWorkflowRunFailedJobsMock.mockResolvedValue([]);

      const testMsg = "Test Message";
      await handleActionFail(testMsg, 0);

      // Behaviour
      expect(setFailedSpy).toHaveBeenCalled();
      expect(setOutputSpy).not.toHaveBeenCalled();

      // Logging
      assertOnlyCalled(coreErrorLogMock);
      expect(coreErrorLogMock).toHaveBeenCalledOnce();
      expect(coreErrorLogMock.mock.calls[0]?.[0]).toMatchInlineSnapshot(
        `"Failed: Test Message"`,
      );
    });

    it("should fetch and log failed jobs from the remote run", async () => {
      const jobs = [
        {
          name: "First Job",
          id: 0,
          status: "completed" as const,
          conclusion: WorkflowRunConclusion.Failure,
          url: "url",
          steps: [
            {
              name: "First Step",
              number: 0,
              status: "completed" as const,
              conclusion: WorkflowRunConclusion.Failure,
            },
          ],
        },
        {
          name: "Second Job",
          id: 0,
          status: "completed" as const,
          conclusion: WorkflowRunConclusion.Failure,
          url: "url",
          steps: [
            {
              name: "First Step",
              number: 0,
              status: "completed" as const,
              conclusion: WorkflowRunConclusion.Success,
            },
          ],
        },
      ];
      apiFetchWorkflowRunFailedJobsMock.mockResolvedValue(jobs);

      const testMsg = "Test Message";
      await handleActionFail(testMsg, 0);

      // Logging
      assertOnlyCalled(coreErrorLogMock);
      expect(coreErrorLogMock).toHaveBeenCalledTimes(3);
      expect(coreErrorLogMock.mock.calls[0]?.[0]).toMatchInlineSnapshot(
        `"Failed: Test Message"`,
      );
      expect(coreErrorLogMock.mock.calls[1]?.[0]).toMatchSnapshot();
      expect(coreErrorLogMock.mock.calls[2]?.[0]).toMatchSnapshot();
    });

    it("should only log steps that did not succeed", async () => {
      const jobs = [
        {
          name: "First Job",
          id: 0,
          status: "completed" as const,
          conclusion: WorkflowRunConclusion.Failure,
          url: "url",
          steps: [
            {
              name: "First Step",
              number: 0,
              status: "completed" as const,
              conclusion: WorkflowRunConclusion.Success,
            },
            {
              name: "Second Step",
              number: 1,
              status: "completed" as const,
              conclusion: WorkflowRunConclusion.Failure,
            },
            {
              name: "Third Step",
              number: 2,
              status: "completed" as const,
              conclusion: WorkflowRunConclusion.Skipped,
            },
          ],
        },
      ];
      apiFetchWorkflowRunFailedJobsMock.mockResolvedValue(jobs);

      const testMsg = "Test Message";
      await handleActionFail(testMsg, 0);

      // Logging
      assertOnlyCalled(coreErrorLogMock);
      expect(coreErrorLogMock).toHaveBeenCalledTimes(2);
      expect(coreErrorLogMock.mock.calls[0]?.[0]).toMatchInlineSnapshot(
        `"Failed: Test Message"`,
      );
      expect(coreErrorLogMock.mock.calls[1]?.[0]).toMatchSnapshot();
    });
  });
});
