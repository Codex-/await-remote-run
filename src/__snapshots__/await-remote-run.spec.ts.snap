// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`await-remote-run > handleActionFail > should fetch and log failed jobs from the remote run 2`] = `
"Job First Job:
  ID: 0
  Status: completed
  Conclusion: failure
  URL: url
  Steps (non-success):
    0: First Step
      Status: completed
      Conclusion: failure"
`;

exports[`await-remote-run > handleActionFail > should fetch and log failed jobs from the remote run 3`] = `
"Job Second Job:
  ID: 0
  Status: completed
  Conclusion: failure
  URL: url
  Steps (non-success):
"
`;

exports[`await-remote-run > handleActionFail > should only log steps that did not succeed 2`] = `
"Job First Job:
  ID: 0
  Status: completed
  Conclusion: failure
  URL: url
  Steps (non-success):
    1: Second Step
      Status: completed
      Conclusion: failure
    2: Third Step
      Status: completed
      Conclusion: skipped"
`;
